### PR TITLE
[webui] Show what depends on information

### PR DESCRIPTION
**Pull request by ChrisBr**
on live log view to determine to trigger a rebuild or not.
#2130.

whatdependson

### CODE REVIEW

**Comment by ChrisBr** [comment]
With a lot of dependencies ...
whatdependson_many

Another possibility to display could be a popup when you click "trigger rebuild"...

**Comment by bgeuken** [comment]
Can't say for sure, but the list of packages might be very long for certain packages. Maybe it makes sense to only show a subset if a certain limit get's exceeded. And maybe a 'more' link that opens a dialog, assuming this is important information.

**Comment by hennevogel** [comment]
If you rebuild this package, it will also trigger rebuilds of 6 other packages. More information[+]

File: src/api/spec/models/package_spec.rb [hunk 1]
```diff
@@ -356,4 +356,78 @@
...
+          "</package>" +
+        "</builddepinfo>"
+      end
+
+      before do
+        stub_request(:get, url).and_return(body: two_dependency)
+        @result = package.what_depends_on(repository, architecture)
+      end
+
+      it 'returns an empty array' do
```
Comment (DavidKang) [inline src/api/spec/models/package_spec.rb]: It's not empty 😄

Comment (ChrisBr) [inline src/api/spec/models/package_spec.rb]: Argh, you're right!

**Comment by ChrisBr** [comment]
whatdependson_hidden

whatdependson_expanded

**Comment by ChrisBr** [comment]
@DavidKang fixed tests!

@hennevogel @bgeuken changed the text

File: src/api/app/models/package.rb [hunk 2]
```diff
@@ -1379,6 +1379,15 @@ def file_exists?(filename)
   def has_icon?
     file_exists?("_icon")
   end
+
+  def what_depends_on(repository, architecture)
+    begin
+      path = "/build/#{project.name}/#{repository}/#{architecture}/_builddepinfo?package=#{name}&view=revpkgnames"
+      [Xmlhash.parse(Suse::Backend.get(path).body).try(:[], 'package').try(:[], 'pkgdep')].flatten.compact
+    rescue ActiveXML::Transport::Error
```
Comment (evanrolfe) [inline src/api/app/models/package.rb]: Do we want to rescue all `ActiveXML::Transport::Error` errors? What if its something important going wrong that we want to know about?

File: src/api/spec/models/package_spec.rb [hunk 3]
```diff
@@ -356,4 +356,78 @@
...
+      package.what_depends_on(repository, architecture)
+      expect(a_request(:get, url)).to have_been_made.once
+    end
+
+    context 'with no build dependencies' do
+      let(:no_dependency) { "<builddepinfo />" }
+
+      before do
+        stub_request(:get, url).and_return(body: no_dependency)
+        @result = package.what_depends_on(repository, architecture)
```
Comment (evanrolfe) [inline src/api/spec/models/package_spec.rb]: Does the team have a preference to use instances variables or let blocks? It seems like let is more commonly used in the codebase and its the preferred way in the [rspec style guidlines](http://betterspecs.org/#let).

**Comment by adrianschroeter** [comment]
I would not write "it will also trigger rebuilds", because these are only the direct depending packages, but there are more indirect one maybe.

On the other side, it might not trigger them depending on the result of build-compare or non-default scheduling strategies.

So, I would just write "packages with direct depend on this package are ..."

File: src/api/app/models/package.rb (hunk 2 above)
Comment (ChrisBr) [inline src/api/app/models/package.rb]: I changed it to ``ActiveXML::Transport::NotFoundError``, this is the only thing what can happen atm when you request for wrong repository or architecture.

File: src/api/spec/models/package_spec.rb (hunk 3 above)
Comment (ChrisBr) [inline src/api/spec/models/package_spec.rb]: I changed it to use let

File: src/api/app/models/package.rb (hunk 2 above)
Comment (evanrolfe) [inline src/api/app/models/package.rb]: nice :+1:

**Comment by ChrisBr** [comment]
@adrianschroeter hm, do you think it makes it easier for the user to understand?
What about

> A rebuild of this package could trigger rebuilds of at least 6 dependent packages. More information[+]

**Comment by bgeuken** [comment]
"could trigger rebuilds" or "might trigger rebuilds".

I would consider putting any additional information to our docu and point users there, like having an icon that links to the article.

**Comment by adrianschroeter** [comment]
On Mittwoch, 15. März 2017, 02:39:06 CET wrote Björn Geuken:
> "could trigger rebuilds" or "might trigger rebuilds".
> 
> I would consider putting any additional information to our docu and point users there, like having an icon that links to the article.


"might trigger rebuilds of these or more" would be the right term ...

But is this still saying something beside "Something might happen" ?

I would not avoid the term "direct dependency" here, because this is all
what we can say for sure.

-- 

Adrian Schroeter
email: adrian@suse.de

SUSE Linux GmbH, GF: Felix Imendörffer, Jane Smithard, Graham Norton, HRB 21284 (AG Nürnberg) 
 
Maxfeldstraße 5                         
90409 Nürnberg 
Germany

**Comment by ChrisBr** [comment]
@bgeuken @hennevogel @adrianschroeter can we agree now on a sentence? I wouldn't overengineer it ...

**Comment by bgeuken** [comment]
As said, I am fine with the proposed sentence. And in case that we want to inform users about the whole concept / process of build dependencies, etc. we should add a pointer to our documentation.

File: src/api/spec/models/package_spec.rb [hunk 4]
```diff
@@ -356,4 +356,91 @@
...
+  describe '#what_depends_on' do
+    let(:repository) { 'openSUSE_Leap_42.1'}
+    let(:architecture) { 'x86_64' }
+    let(:parameter) { "package=#{package.name}&view=revpkgnames" }
+    let(:url) { "#{CONFIG['source_url']}/build/#{package.project}/#{repository}/#{architecture}/_builddepinfo?#{parameter}" }
+    let(:result) { Package.what_depends_on(package.project, package, repository, architecture) }
+    let(:no_dependency) { "<builddepinfo />" }
+
+    it "should call 'bar' with appropriate arguments" do
+      stub_request(:get, url).and_return(body: no_dependency)
```
Comment (bgeuken) [inline src/api/spec/models/package_spec.rb]: Do we actually need this here? I don't think Package.what_depends_on will get called in this test

File: src/api/spec/models/package_spec.rb [hunk 5]
```diff
@@ -356,4 +356,91 @@
...
+    end
+  end
+
+  describe '#what_depends_on' do
+    let(:repository) { 'openSUSE_Leap_42.1'}
+    let(:architecture) { 'x86_64' }
+    let(:parameter) { "package=#{package.name}&view=revpkgnames" }
+    let(:url) { "#{CONFIG['source_url']}/build/#{package.project}/#{repository}/#{architecture}/_builddepinfo?#{parameter}" }
+    let(:result) { Package.what_depends_on(package.project, package, repository, architecture) }
+    let(:no_dependency) { "<builddepinfo />" }
```
Comment (bgeuken) [inline src/api/spec/models/package_spec.rb]: IMO you only need repository and architecture here

File: src/api/spec/models/package_spec.rb [hunk 6]
```diff
@@ -356,4 +356,91 @@
...
+        stub_request(:get, url).and_return(body: one_dependency)
+      end
+
+      it 'returns an array with the dependency' do
+        expect(result).to eq(['gcc'])
+      end
+    end
+
+    context 'with more than one build dependency' do
+      let(:two_dependency) do
```
Comment (bgeuken) [inline src/api/spec/models/package_spec.rb]: two_dependencies?:-)

File: src/api/app/models/package.rb [hunk 7]
```diff
@@ -1379,6 +1379,19 @@ def file_exists?(filename)
...
+      path = "/build/#{project}/#{repository}/#{architecture}/_builddepinfo?package=#{package}&view=revpkgnames"
+      [Xmlhash.parse(Suse::Backend.get(path).body).try(:[], 'package').try(:[], 'pkgdep')].flatten.compact
+    rescue ActiveXML::Transport::NotFoundError
+      []
+    end
+  end
+
+  def what_depends_on(repository, architecture)
+    Package.what_depends_on(project.name, name, repository, architecture)
+  end
```
Comment (bgeuken) [inline src/api/app/models/package.rb]: If we don't use this method, I don't think it's worth implementing it

**Comment by ChrisBr** [comment]
@bgeuken fixed your suggestions!

**Comment by bgeuken** [comment]
LGTM

**Pull request merged**